patterns/interpreter.md

```diff
@@ -0,0 +1,317 @@
+# Interpreter
+
+## Description
+
+If a problem occurs very often and requires long and repetitive steps to solve it,
+then the problem instances might be expressed in a simple
+language and an interpreter object could solve it by
+interpreting the sentences written in this simple language.
+Basically, for any kind of problems we define a domain
+language, then define a grammar for this language and
+design interpreter solving problem instances.
+
+## Motivation
+
+Imagine that our work is translating simple mathematical expressions into
+[assembly language](https://en.wikipedia.org/wiki/Assembly_language)
+(more simple and low level programming language).
+For simplicity, our expressions consists of ten digits `0`,...,`9`,
+four operations `+, -, /, *` and a pair of parenthesis `(, )`.
+For example, expression `2 + 4` could be translated into
+
+```ignore
+mov eax, 2
+mov ebx, 4
+add eax, ebx
+```
+
+Our goal is to automate translating into assembly instructions
+using the Interpreter design pattern. In other words, we want simply
+provide the Interpreter with an expression and get Assembly
+language output. For example
+
+```rust, ignore
+x.interpret("7+3*(2-1)", &output);
+```
+
+## Context Free Grammars
+
+Basicaly, a
+[Context Free Grammar](https://en.wikipedia.org/wiki/Context-free_grammar)
+(CFG) describes (or defines) all possible strings
+over some finite set of symbols, formally called alphabet.
+These strings usually has some well defined structure.
+For example, strings of all balanced parentheses
+
+```ignore
+(), (()), ()(), ()(()),..
+```
+
+Formally a CFG consists of three sets: terminal symbols,
+nonterminal symbols, a start (nonterminal) symbol,
+and production rules.
+
+In the previous example we have two terminal symbols `(` and `)`.
+We define a single nonterminal (at the same time the start)
+symbol `S` and priduction rules as
+
+```ignore
+S -> SS
+S ->(S)
+S ->()
+```
+
+So, in order to generate a string with balanced parentheses we
+repeatedly apply production rules.
+For example, to derive `()()` we apply the following rules
+
+```ignore
+S -> SS -> ()(S) -> ()()
+```
+
```

Comment:
Here we have to start a new section, just to make it clear that the previous example is finished.

Reply:
Actually, I think we should move this section, because it is in the middle of the assembly stuff. I would put this before the assembly discussion. What do you think?

Reply:
@MarcoIeni I don't mind. Could you update yourself?